Meta/Lagom/Tools/CodeGenerators/LibWeb/GenerateWebGLRenderingContext.cpp

```diff
@@ -713,17 +713,38 @@ class @class_name@ {
             }
             function_impl_generator.set("number_of_vector_elements", number_of_vector_elements);
             function_impl_generator.append(R"~~~(
+    @cpp_element_type@ const* data = nullptr;
+    size_t count = 0;
     if (v.has<Vector<@cpp_element_type@>>()) {
-        auto& data = v.get<Vector<@cpp_element_type@>>();
-        glUniform@number_of_vector_elements@@gl_postfix@v(location->handle(), data.size() / @number_of_vector_elements@, data.data());
-        return;
+        auto& vector = v.get<Vector<@cpp_element_type@>>();
+        data = vector.data();
+        count = vector.size();
+    } else if (v.has<GC::Root<WebIDL::BufferSource>>()) {
+        auto& typed_array_base = static_cast<JS::TypedArrayBase&>(*v.get<GC::Root<WebIDL::BufferSource>>()->raw_object());
+        auto& typed_array = verify_cast<JS::@typed_array_type@>(typed_array_base);
+        data = typed_array.data().data();
+        count = typed_array.array_length().length();
+    } else {
+        VERIFY_NOT_REACHED();
+    }
+)~~~");
+
+            // FIXME: If srcOffset + srcLength is longer than data.length, generate INVALID_VALUE.
+            if (webgl_version == 2) {
+                function_impl_generator.append(R"~~~(
+    data += src_offset;
+    if (src_length == 0) {
+        count -= src_offset;
+    }
+
+    if (src_offset + src_length <= count) {
+        set_error(GL_INVALID_VALUE);
```

Comment:
should we skip the gl call if we generate an invalid value error? (i.e. return early).

also if we are generating the error, probably dont' need the fixme on 732

Reply:
ah, right, I forgot to add early return here 🤦🏻 fixed!